README.md

```diff
@@ -22,14 +30,83 @@ dependencies: [
+make sure to feed it through the entire system. At cross-cutting boundaries, e.g. right before sending an HTTP
+request, they'd inject the `BaggageContext` into the HTTP headers, allowing context propagation. On the flip-sight, an
+HTTP server should extract the request headers into a `BaggageContext`. Injecting/extracting is part of the
+`swift-tracing` libraries [and documented in its own repository](https://github.com/slashmo/gsoc-swift-tracing).
+
+### Instrumentation
+
+When implementing instrumentation for cross-cutting tools, `BaggageContext` becomes the way you propagate metadata such
+as trace ids. Because each instrument knows what values might be added to the `BaggageContext` they are the ones
+creating `BaggageContextKey` types dictating the type of value associated with each key added to the context:
+
+```swift
+private enum TraceIDKey: BaggageContextKey {
+  typealias Value = String
+}
+
+var context = BaggageContext()
+context[TraceIDKey.self] = "4bf92f3577b34da6a3ce929d0e0e4736"
+print(context[TraceIDKey.self] ?? "new trace id")
+```
+
+**TODO**: Add paragraph on why keys should be private/internal (#25)
+
+To make accessing values a bit more convenient, we encourage you to add computed properties to `BaggageContextProtocol`:
+
+```swift
+extension BaggageContextProtocol {
+  var traceID: String? {
+    get {
+      return self[TraceIDKey.self]
+    }
+    set {
+      self[TraceIDKey.self] = newValue
+    }
+  }
+}
+```
+
+### Distributed System
+
+When creating distributed systems, each of your components will likely make use of one or more generically instrumented
+frameworks/libraries. This means you get to choose what instrument(s) to use 🙌. Check out
+[the swift-tracing](https://github.com/slashmo/gsoc-swift-tracing) for instructions on how to get up & running.
+
+Occasionally, when using multiple frameworks/libraries together, you'll need to pass along a `BaggageContext` to ensure
+that the values stored in this context get propagated correctly.
+
 ## Context-Passing Guidelines
 
-In order for context-passing to feel consistent and Swifty among all server-side (and not only) libraries and frameworks
-aiming to adopt `BaggageContext` (or any of its uses, such as Distributed Tracing), we suggest the following set of guidelines:
+For context-passing to feel consistent and Swifty among all server-side (and not only) libraries and frameworks
+aiming to adopt `BaggageContext` (or any of its uses, such as Distributed Tracing), we suggest the following set of
+guidelines:
 
 ### Argument naming/positioning
 
-In order to propagate baggage through function calls (and asynchronous-boundaries it may often be necessary to pass it explicitly (unless wrapper APIs are provided which handle the propagation automatically).
+To propagate baggage through function calls (and asynchronous-boundaries) it may often be necessary to pass it
```

Comment:
```suggestion
Propagating baggage context through your system is to be done explicitly, meaning as a parameter in function calls, following the "flow" of execution. 

> Do not store the context object in your classes or structs as this will inevitably lead to mistakes and hard to track bugs.
```